rust/catalyst-types/src/id_uri/mod.rs

```diff
@@ -0,0 +1,533 @@
+    /// time. If a URI does not have a defined nonce, this function will always return
+    /// `false`.
+    ///
+    /// # Arguments
+    ///
+    /// * `self`: A reference to the URI object that contains the potential nonce.
+    /// * `past`: The duration by which we look back in time from the current moment
+    ///   (`now`). (Positive Duration)
+    /// * `future`: The duration by which we look forward in time from the current moment
+    ///   (`now`). (Positive Duration)
+    ///
+    /// # Returns
+    ///
+    /// A boolean value:
+    /// - `true` if the nonce is within the specified range relative to `now()`.
+    /// - `false` if there is no nonce defined or if the nonce falls outside the specified
+    ///   range.
+    ///
+    /// If the URI does not have a nonce defined, this function returns `false`
+    /// immediately because it cannot perform the check without a valid nonce present.
+    /// This behavior ensures that the absence of a nonce will fail any required range
+    /// checks when such checks are expected according to the function's contract.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use catalyst_types::id_uri::IdUri;
+    /// use chrono::{DateTime, Duration, Utc};
+    /// let uri = "id.catalyst://cardano/FftxFnOrj2qmTuB2oZG2v0YEWJfKvQ9Gg8AgNAhDsKE"
+    ///     .parse::<IdUri>()
+    ///     .unwrap()
+    ///     .with_nonce();
+    /// // true, within range
+    /// assert!(uri.is_nonce_in_range(chrono::Duration::hours(1), chrono::Duration::minutes(5)));
+    ///
+    /// // Change the nonce to be 1970/1/1 00:00:00
+    /// let uri = uri.with_specific_nonce(DateTime::<Utc>::MIN_UTC);
+    /// // false, outside range
+    /// assert!(!uri.is_nonce_in_range(chrono::Duration::hours(1), chrono::Duration::minutes(5)));
+    /// ```
+    #[must_use]
+    pub fn is_nonce_in_range(&self, past: Duration, future: Duration) -> bool {
+        if let Some(nonce) = self.nonce {
+            let now = Utc::now();
+            let start_time = now - past;
+            let end_time = now + future;
+            (start_time..=end_time).contains(&nonce)
+        } else {
+            // No nonce defined, so we say that this fails.
+            // Prevents an absent Nonce from passing range checks when its required.
+            false
+        }
+    }
+
+    /// Convert the `IdUri` to its shortest form.
+    /// This is an ID without any role/rotation information, no scheme, no username or
+    /// nonce.
+    /// This is used to get the most generalized form of a Catalyst ID.
+    #[must_use]
+    pub fn as_short_id(&self) -> Self {
+        self.clone()
+            .with_role(RoleIndex::default())
+            .with_rotation(KeyRotation::default())
+            .without_username()
+            .without_nonce()
+            .without_encryption()
+            .as_id()
+    }
+}
+
+impl FromStr for IdUri {
+    type Err = errors::IdUriError;
```

Comment:
Do we want to use a `ProblemReport` at this place ?

Reply:
No, it fails once, and doesn't continue.
ProblemReports should be for things that are user generated and are much more complex than a URI.
However, the error from this may end up in a Problem Report if this type is in such a struct.